### PR TITLE
Trace and shrink dictionaries as TypedDicts if possible.

### PR DESCRIPTION
**Pull request by pradeep90**
* Trace a dictionary as an anonymous TypedDict if all keys are strings (up to a limit).
* Conservatively shrink multiple TypedDicts as one only if they are all identical. Otherwise, fall back to the default behaviour.
* Generate a TypedDict class definition stub at the top of the module stub and name it using the parameter name.
* Add MAX_TYPED_DICT_SIZE field to Config.
* Hand-roll an equality test because the default equality test for TypedDict fails for identical invocations: TypedDict('Foo', {'a': int}) != TypedDict('Foo', {'a': int}).
* Trace and shrink return types and yield types as well.
* Handle nested dictionaries.

* Add `make_forward_ref` so that we can use ForwardRefs. I deleted a test for a nested `Dict[str, TypedDict]` because it kept failing on Python 3.6.

### CODE REVIEW

File: monkeytype/typing.py [hunk 1]
```diff
@@ -52,7 +114,19 @@ def make_generator(yield_typ, send_typ, return_typ):
 )
 
 
-def get_type(obj):
+def get_dict_type(dct, max_num_typed_dict_keys=None):
```
Comment (carljm) [inline monkeytype/typing.py]: nit: I prefer the slightly shorter `max_typed_dict_size`

File: monkeytype/typing.py [hunk 2]
```diff
@@ -20,19 +20,81 @@
...
-    """Return the smallest type equivalent to Union[types]"""
+    """Return the smallest type equivalent to Union[types].
+    If all the types are TypedDicts, shrink them ourselves.
+    Otherwise, turn the TypedDicts into Dicts. Union will handle deduplicating
+    types (both by equality and subtype relationships)."""
     types = tuple(types)
     if len(types) == 0:
         return Any
-    # Union will handle deduplicating types (both by equality and subtype
-    # relationships)
```
[the remark this answers is not in the document]
Comment (pradeep90) [reply monkeytype/typing.py]: I had moved it to the docstring.

File: tests/test_typing.py [hunk 3]
```diff
@@ -24,16 +24,61 @@
...
+                TypedDict(DUMMY_TYPED_DICT_NAME, {'a': int, 'b': int}),
+            ),
+            # Extra key in one dictionary - fall back to Union, which gives Dict[str, int].
+            (
+                (
+                    TypedDict(DUMMY_TYPED_DICT_NAME, {'a': int, 'b': int}),
+                    TypedDict(DUMMY_TYPED_DICT_NAME, {'a': int}),
+                ),
+                Dict[str, int],
+            ),
```
[the remark this answers is not in the document]
Comment (pradeep90) [reply tests/test_typing.py]: Yeah, I wanted to check how many use cases were covered with the simplest case and add other shrinking strategies later.

**Comment by federicobond** [comment]
Hi there, I am encountering an error potentially related to this behaviour:
> Generate a TypedDict class definition stub at the top of the module stub and name it using the parameter name.

If I have a module like this:
```python
def foo():
    return dict(a=1, b='str')
```
It generates a stub file like this:
```python
from mypy_extensions import TypedDict


class FooTypedDict(TypedDict):
    a: int
    b: str
```

This stub does not apply to the original module:
```
error: [...].py: Class 'FooTypedDict' not found in source.
```

**Comment by pradeep90** [comment]
@carljm Could this be because the generated class definition `FooTypedDict` is present only in the `.pyi` file but not in the `.py` file?

@federicobond I'm guessing the error goes away after you copy-paste the class definition to the source file?

**Comment by carljm** [comment]
@pradeep90 Yes, I expect that's the source of the error :/ I think I will need to make a new release of MonkeyType with the TypedDict feature turned off by default, until we are able to resolve the issue (e.g. by adding support to retype to copy entire TypedDict classes from stub to code, or with some other workaround.)

**Comment by carljm** [comment]
https://github.com/Instagram/MonkeyType/pull/155

**Comment by carljm** [comment]
Thanks @federicobond for reporting this!

**Pull request merged**